src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
```

Comment:
Tried it out and we realized this shouldn't be done lightly (it affects other customers' use of `INSERT` within their own functions, many of which presently hit multiple shards within a single function).